extension/extensionlimiter/rate.go

```diff
@@ -0,0 +1,95 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package extensionlimiter // import "go.opentelemetry.io/collector/extension/extensionlimiter"
+
+import (
+	"context"
+)
+
+// RateLimiterProvider is a provider for rate limiters.
+//
+// Limiter implementations will implement this or the
+// ResourceLimiterProvider interface, but MUST not implement both.
+// Limiters are covered by configmiddleware configuration, which is
+// able to construct LimiterWrappers from these providers.
+type RateLimiterProvider interface {
+	// RateLimiter returns a provider for rate limiters.
+	RateLimiter(WeightKey, ...Option) (RateLimiter, error)
+}
+
+// RateLimiterProviderFunc is a functional way to build RateLimters.
+type RateLimiterProviderFunc func(WeightKey, ...Option) (RateLimiter, error)
+
+var _ RateLimiterProvider = RateLimiterProviderFunc(nil)
+
+// RateLimiter implements RateLimiterProvider.
+func (f RateLimiterProviderFunc) RateLimiter(key WeightKey, opts ...Option) (RateLimiter, error) {
+	return f(key, opts...)
+}
+
+// RateLimiter is an interface that an implementation makes available
+// to apply time-based limits on quantities such as the number of
+// bytes or items per second.
+//
+// This is a relatively low-level interface. Callers that can use a
+// LimiterWrapper should choose that interface instead. This interface
+// is meant for direct use only in special cases where control flow
+// cannot be easily scoped to a callback, for example inside
+// middleware (e.g., grpc.StatsHandler).
+//
+// See the README for more recommendations.
+type RateLimiter interface {
+	// Limiter includes MustDeny().
+	Limiter
+
```

Comment:
here is my confusion, I think the `RateLimiterProvider` should implement the Limiter, because I want to call the "saturation" only once asap. I know I can still do that by checking all the weight type, but for a limiter that supports only items I don't think you expect to call this in the "middleware".

Reply:
`RateLimiterProvider` is called in Start() typically (for static limiter configurations). Are we on the same page?

Re: only once: I created an adapter for combining multiple limiters (which may be called at different times) by their MustDeny(). That's how I call MustDeny once, it's the `MultiLimiter` in limiterhelper/consumer.go. When a wrapper limiter constructs >1 limiter in the wrapper, there is a return value limiter which combines all of them for MustDeny calls.

Reply:
Update: To combine `Checker` instances into a single instance, use `limiterhelper.MultiChecker`. The three basic Checker constructors are NewLimiterWrapperChecker, NewRateLimiterChecker, and NewResourceLimiterChecker.

Reply:
I think I sorted this out. Thank you for pointing to this. I have modified the RateLimiter and ResourceLimiter interfaces to include GetChecker() methods which allow obtaining a Checker, having the MustDeny method. The MustDeny method and Checker interface apply to all weight keys. I have documented that MustDeny may be called more than once per request since middleware, receiver, and wrappers can all reasonably call this in various circumstances.